File: src/test/java/com/gemini/krakenbot/service/PortfolioManagerDrawdownTest.java
package com.gemini.krakenbot.service;

import com.gemini.krakenbot.service.impl.*;

import com.gemini.krakenbot.config.Allocation;
import com.gemini.krakenbot.config.AppConfig;
import com.gemini.krakenbot.config.Settings;
import com.gemini.krakenbot.model.PortfolioSnapshot;
import com.gemini.krakenbot.model.PortfolioStats;
import com.gemini.krakenbot.repository.PortfolioStatsRepository;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.mockito.ArgumentCaptor;
import org.mockito.InjectMocks;
import org.mockito.Mock;
import org.mockito.MockitoAnnotations;
import org.springframework.test.util.ReflectionTestUtils;

import java.math.BigDecimal;
import java.util.Collections;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.mockito.Mockito.*;

class PortfolioManagerDrawdownTest {

    @Mock
    private KrakenService krakenService;
    @Mock
    private ConfigService configService;
    @Mock
    private TradeHistoryService tradeHistoryService;
    @Mock
    private PortfolioStatsRepository portfolioStatsRepository;
    @Mock
    private AppConfig appConfig;
    @Mock
    private Settings settings;

    @InjectMocks
    private PortfolioManagerImpl portfolioManager;

    @BeforeEach
    void setUp() {
        MockitoAnnotations.openMocks(this);
        when(configService.getConfig()).thenReturn(appConfig);
        when(appConfig.settings()).thenReturn(settings);

        // Default settings
        when(settings.loopDelaySeconds()).thenReturn(1L);
        when(settings.dryRun()).thenReturn(false);
        when(settings.deviationTriggerPercent()).thenReturn(2.0);
        when(settings.dustThresholdUSD()).thenReturn(1.0);
        when(settings.fiatMaxDrawdown()).thenReturn(50.0); // 50% Max Drawdown
        when(settings.fiatDeploymentExponent()).thenReturn(1.0); // Linear
    }

    @Test
    void testDrawdownAndFiatDeployment() {
        // 1. Setup Portfolio State
        // ATH = $2000
        // Current = $1000 (50% Drawdown)
        // With MaxDrawdown = 50%, we expect 100% Deployment.
        // Or let's do 25% Drawdown (Current $1500) -> 50% ratio -> 50% Deployment.

        when(portfolioStatsRepository.load()).thenReturn(new PortfolioStats(new BigDecimal("2000.0")));

        // Allocations: A=50%, USD=50%
        List<Allocation> allocs = List.of(
                new Allocation("A", 50.0),
                new Allocation("USD", 50.0));
        when(appConfig.allocations()).thenReturn(allocs);

        // Prices
        Map<String, Double> prices = new HashMap<>();
        prices.put("AUSD", 100.0);
        when(krakenService.getTickerPrices(anyString())).thenReturn(prices);

        // Balances:
        // A = 7.5 units ($750)
        // USD = $750
        // Total = $1500.
        // Drawdown = (2000 - 1500) / 2000 = 0.25 (25%).
        // MaxDD = 50. Ratio = 25/50 = 0.5.
        // Exponent = 1.0. Deploy = 0.5^1 * 100 = 50%.
        // Effective USD Target:
        // Base USD Target = 50%.
        // Factor = 1 - (50/100) = 0.5.
        // Effective USD Target = 50% * 0.5 = 25%.
        // Effective Crypto Target (A) = 100% - 25% = 75%?
        // No, logic in code only adjusts USD target. It DOES NOT automatically boost
        // others.
        // But target check calculates deviation.
        // USD Target Value = Total($1500) * 25% = $375.
        // Current USD = $750.
        // Deviation = $750 - $375 = +$375 (Overweight).
        // A Target Value = Total($1500) * 50% = $750.
        // Current A = $750.
        // Deviation = 0.
        // Logic: USD is +$375 overweight. It should trigger a BUY of something?
        // Code logic: "USD Triggered but no Crypto Triggered" ->
        // distributeFiatCorrection.
        // Note: A's deviation % is 0. So it won't be in buy/sell orders.
        // distributeFiatCorrection should be called.

        Map<String, Double> balances = new HashMap<>();
        balances.put("A", 7.5);
        balances.put("USD", 750.0);
        when(krakenService.getBalances()).thenReturn(balances);

        // Act
        ReflectionTestUtils.invokeMethod(portfolioManager, "performRebalanceCycle");

        // Assert
        // We expect USD to be considered overweight by $375, and that surplus to be put
        // into A.
        // A is the only candidate.
        // Executing BUY A for $375.
        // Vol = 375 / 100 = 3.75.

        verify(krakenService).executeOrder(eq("AUSD"), eq("market"), eq("buy"),
                doubleThat(v -> Math.abs(v - 3.75) < 0.01));

        // Also verify snapshot recorded values
        ArgumentCaptor<PortfolioSnapshot> captor = ArgumentCaptor.forClass(PortfolioSnapshot.class);
        verify(tradeHistoryService).addSnapshot(captor.capture());
        PortfolioSnapshot s = captor.getValue();

        assertEquals(25.0, s.getDrawdownPercent().doubleValue(), 0.01);
        assertEquals(50.0, s.getFiatDeploymentPercent().doubleValue(), 0.01);
        assertEquals(25.0, s.getEffectiveUsdTargetPercent().doubleValue(), 0.01);
    }

    @Test
    void testNewATH() {
        // Previous ATH = 1000. Current = 1500.
        // Should update ATH.
        PortfolioStats stats = new PortfolioStats(new BigDecimal("1000.0"));
        when(portfolioStatsRepository.load()).thenReturn(stats);

        List<Allocation> allocs = List.of(new Allocation("USD", 100.0));
        when(appConfig.allocations()).thenReturn(allocs);
        when(krakenService.getTickerPrices(anyString())).thenReturn(Collections.emptyMap());

        Map<String, Double> balances = new HashMap<>();
        balances.put("USD", 1500.0);
        when(krakenService.getBalances()).thenReturn(balances);

        ReflectionTestUtils.invokeMethod(portfolioManager, "performRebalanceCycle");

        verify(portfolioStatsRepository).save(stats);
        assertEquals(0, new BigDecimal("1500.0").compareTo(stats.getAllTimeHigh()));
    }
}

File: src/test/kotlin/com/gemini/krakenbot/service/PortfolioManagerDrawdownTest.kt
package com.gemini.krakenbot.service

import com.fasterxml.jackson.module.kotlin.jacksonObjectMapper
import com.gemini.krakenbot.TestFixtures
import com.gemini.krakenbot.config.Allocation
import com.gemini.krakenbot.config.AppConfig
import com.gemini.krakenbot.config.DatabaseConfig
import com.gemini.krakenbot.config.KrakenCredentials
import com.gemini.krakenbot.model.Asset
import com.gemini.krakenbot.model.PortfolioSnapshot
import com.gemini.krakenbot.model.PortfolioStats
import com.gemini.krakenbot.repository.PortfolioStatsRepository
import com.gemini.krakenbot.repository.impl.SqlitePortfolioStatsRepositoryImpl
import com.gemini.krakenbot.repository.table.PortfolioStatsTable
import com.gemini.krakenbot.service.impl.OrderExecutorImpl
import com.gemini.krakenbot.service.impl.PortfolioAnalyzerImpl
import com.gemini.krakenbot.service.impl.PortfolioManagerImpl
import io.kotest.assertions.throwables.shouldThrow
import io.kotest.core.spec.IsolationMode
import io.kotest.core.spec.style.StringSpec
import io.kotest.matchers.booleans.shouldBeTrue
import io.kotest.matchers.comparables.shouldBeEqualComparingTo
import io.kotest.matchers.nulls.shouldNotBeNull
import io.kotest.matchers.shouldBe
import io.mockk.coEvery
import io.mockk.coVerify
import io.mockk.every
import io.mockk.mockk
import io.mockk.slot
import kotlinx.coroutines.test.runTest
import org.jetbrains.exposed.v1.jdbc.selectAll
import org.jetbrains.exposed.v1.jdbc.transactions.transaction
import java.io.File
import java.io.IOException
import java.math.BigDecimal
import java.math.RoundingMode

class PortfolioManagerDrawdownTest : StringSpec() {

    override fun isolationMode() = IsolationMode.InstancePerTest

    private val krakenService = FakeKrakenService()
    private val configService = mockk<ConfigService>(relaxed = true)
    private val tradeHistoryService = mockk<TradeHistoryService>(relaxed = true)
    private val portfolioStatsRepository =
        mockk<PortfolioStatsRepository>(relaxed = true)
    private lateinit var portfolioManager: PortfolioManagerImpl
    private lateinit var portfolioAnalyzer: PortfolioAnalyzer
    private lateinit var orderExecutor: OrderExecutor

    init {
        beforeTest {
            portfolioAnalyzer = PortfolioAnalyzerImpl(
                krakenService = krakenService,
                configService = configService,
                portfolioStatsRepository = portfolioStatsRepository,
            )
            orderExecutor = OrderExecutorImpl(krakenService, tradeHistoryService)
            portfolioManager = PortfolioManagerImpl(
                configService = configService,
                tradeHistoryService = tradeHistoryService,
                portfolioAnalyzer = portfolioAnalyzer,
                orderExecutor = orderExecutor,
            )

            val settings = TestFixtures.settings(dryRun = false, loopDelaySeconds = 60L, fiatMaxDrawdown = 50.0)
            val appConfig =
                AppConfig(
                    kraken = KrakenCredentials(
                        apiKey = "k",
                        privateKey = "s",
                    ),
                    settings = settings,
                    allocations = emptyList(),
                )
            every { configService.getConfig() } returns appConfig
        }

        "testDrawdownAndFiatDeployment" {
            runTest {
                coEvery {
                    portfolioStatsRepository.load()
                } returns PortfolioStats(
                    BigDecimal("2000.0"),
                )

                val allocs = listOf(
                    Allocation("A", 50.0),
                    Allocation(Asset.USD, 50.0),
                )

                val appConfig = TestFixtures.config(
                    settings = TestFixtures.settings(dryRun = false, loopDelaySeconds = 60L, fiatMaxDrawdown = 50.0),
                    allocations = allocs,
                )
                every { configService.getConfig() } returns appConfig

                val prices = mapOf("AUSD" to 100.0)
                krakenService.pricesSupplier = { prices }

                val balances = mapOf(
                    "A" to 7.5,
                    Asset.USD to 750.0,
                )
                krakenService.balanceSupplier = { balances }

                portfolioManager.performRebalanceCycle()

                krakenService.executedOrders.size shouldBe 1
                val order = krakenService.executedOrders[0]
                order.pair shouldBe "AUSD"
                order.type shouldBe "market"
                order.side shouldBe "buy"
                (
                    order.volume.subtract(BigDecimal.valueOf(3.75))
                        .abs() < BigDecimal("0.01")
                    ).shouldBeTrue()

                val captor = slot<PortfolioSnapshot>()
                coVerify { tradeHistoryService.addSnapshot(capture(captor)) }
                val s = captor.captured

                s.drawdownPercent.shouldBeEqualComparingTo(BigDecimal("25.0"))
                s.fiatDeploymentPercent.shouldBeEqualComparingTo(BigDecimal("50.0"))
                s.effectiveUsdTargetPercent.shouldBeEqualComparingTo(BigDecimal("25.0"))
            }
        }

        "testNewATH" {
            runTest {
                val stats = PortfolioStats(BigDecimal("1000.0"))
                coEvery { portfolioStatsRepository.load() } returns stats

                val allocs = listOf(
                    Allocation(
                        symbol = Asset.USD,
                        targetPercent = 100.0,
                    ),
                )

                val appConfig = TestFixtures.config(
                    settings = TestFixtures.settings(dryRun = false, loopDelaySeconds = 60L, fiatMaxDrawdown = 50.0),
                    allocations = allocs,
                )
                every { configService.getConfig() } returns appConfig
                krakenService.pricesSupplier = { emptyMap() }

                val balances = mapOf(Asset.USD to 1500.0)
                krakenService.balanceSupplier = { balances }

                portfolioManager.performRebalanceCycle()

                val captor = slot<PortfolioStats>()
                coVerify { portfolioStatsRepository.save(capture(captor)) }
                captor.captured.allTimeHigh.shouldNotBeNull()
                captor.captured.allTimeHigh.shouldBeEqualComparingTo(BigDecimal("1500.0"))
            }
        }

        "USD-only drawdown reports zero deployment and keeps the full USD target" {
            runTest {
                coEvery { portfolioStatsRepository.load() } returns PortfolioStats(BigDecimal("1000.00"))
                every { configService.getConfig() } returns
                    TestFixtures.config(
                        settings =
                        TestFixtures.settings(dryRun = false, loopDelaySeconds = 60L, fiatMaxDrawdown = 50.0),
                        allocations = listOf(Allocation(Asset.USD, 100.0)),
                    )
                krakenService.pricesSupplier = { emptyMap() }
                krakenService.balanceSupplier = { mapOf(Asset.USD to 500.0) }

                portfolioManager.performRebalanceCycle()

                val snapshot = slot<PortfolioSnapshot>()
                coVerify { tradeHistoryService.addSnapshot(capture(snapshot)) }
                snapshot.captured.drawdownPercent.shouldBeEqualComparingTo(BigDecimal("50.0000"))
                snapshot.captured.fiatDeploymentPercent.shouldBeEqualComparingTo(BigDecimal.ZERO)
                snapshot.captured.effectiveUsdTargetPercent.shouldBeEqualComparingTo(BigDecimal("100.0"))
                krakenService.executedOrders.size shouldBe 0
            }
        }

        "corrupt ATH migration aborts analysis before saving a lower ATH or planning orders" {
            runTest {
                val statsFile = File("test-ath-fail-closed-stats.json")
                val statsBackup = File("test-ath-fail-closed-stats.json.bak")
                val isolatedDb = DatabaseConfig.init(TestFixtures.MEMORY_)
                val statsRepository =
                    SqlitePortfolioStatsRepositoryImpl(isolatedDb, jacksonObjectMapper(), statsFile.path)
                val failClosedAnalyzer =
                    PortfolioAnalyzerImpl(
                        krakenService = krakenService,
                        configService = configService,
                        portfolioStatsRepository = statsRepository,
                    )
                val failClosedManager =
                    PortfolioManagerImpl(
                        configService = configService,
                        tradeHistoryService = tradeHistoryService,
                        portfolioAnalyzer = failClosedAnalyzer,
                        orderExecutor = OrderExecutorImpl(krakenService, tradeHistoryService),
                    )
                val config =
                    TestFixtures.config(
                        settings =
                        TestFixtures.settings(dryRun = false, loopDelaySeconds = 60L, fiatMaxDrawdown = 50.0),
                        allocations =
                        listOf(
                            Allocation(Asset.BTC, 50.0),
                            Allocation(Asset.USD, 50.0),
                        ),
                    )
                every { configService.getConfig() } returns config
                krakenService.pricesSupplier = { mapOf(Asset.BTC_USD_PAIR to 100.0) }
                krakenService.balanceSupplier = { mapOf(Asset.BTC to 0.0, Asset.USD to 1000.0) }

                try {
                    statsFile.delete()
                    statsBackup.delete()
                    statsFile.writeText("{not-json")

                    shouldThrow<IOException> {
                        failClosedManager.performRebalanceCycle()
                    }

                    krakenService.executedOrders.size shouldBe 0
                    coVerify(exactly = 0) { tradeHistoryService.addSnapshot(any()) }
                    transaction(isolatedDb) {
                        PortfolioStatsTable.selectAll().count() shouldBe 0
                    }
                    statsFile.exists() shouldBe true
                    statsBackup.exists() shouldBe false
                } finally {
                    statsFile.delete()
                    statsBackup.delete()
                }
            }
        }

        "testCalculateFiatDeployment_AtMaxDrawdownSaturationExponentOne" {
            runTest {
                val settings = TestFixtures.settings(dryRun = false, loopDelaySeconds = 60L, fiatMaxDrawdown = 50.0)
                portfolioAnalyzer.calculateFiatDeployment(
                    BigDecimal("50.0"),
                    settings,
                ).shouldBeEqualComparingTo(BigDecimal("100.0"))
            }
        }

        "testCalculateFiatDeployment_AtMaxDrawdownSaturationExponentTwo" {
            runTest {
                val settings = TestFixtures.settings(
                    dryRun = false,
                    loopDelaySeconds = 60L,
                    fiatMaxDrawdown = 50.0,
                    fiatDeploymentExponent = 2.0,
                )
                portfolioAnalyzer.calculateFiatDeployment(
                    BigDecimal("50.0"),
                    settings,
                ).shouldBeEqualComparingTo(BigDecimal("100.0"))
            }
        }

        "testCalculateFiatDeployment_AboveMaxDrawdownCoercesTo100" {
            runTest {
                val settings = TestFixtures.settings(
                    dryRun = false,
                    loopDelaySeconds = 60L,
                    fiatMaxDrawdown = 50.0,
                    fiatDeploymentExponent = 2.0,
                )
                portfolioAnalyzer.calculateFiatDeployment(
                    BigDecimal("75.0"),
                    settings,
                ).shouldBeEqualComparingTo(BigDecimal("100.0"))
            }
        }

        // CQ-3-18: ALGORITHM.md MaxDD=30% aggressive exponent 0.5 table
        // | DD 1.5% → 22% | 7.5% → 50% | 15% → 71% | 22.5% → 87% | 30% → 100% |
        "testCalculateFiatDeployment_AggressiveExponentHalf_AlgorithmTable" {
            runTest {
                val settings = TestFixtures.settings(
                    dryRun = false,
                    loopDelaySeconds = 60L,
                    fiatMaxDrawdown = 30.0,
                    fiatDeploymentExponent = 0.5,
                )
                // Documented integer Deploy% (ALGORITHM rounds for display).
                val algorithmTable =
                    listOf(
                        BigDecimal("1.5") to BigDecimal("22"),
                        BigDecimal("7.5") to BigDecimal("50"),
                        BigDecimal("15") to BigDecimal("71"),
                        BigDecimal("22.5") to BigDecimal("87"),
                        BigDecimal("30") to BigDecimal("100"),
                    )
                for ((drawdownPct, tableDeployPct) in algorithmTable) {
                    val deploy =
                        portfolioAnalyzer.calculateFiatDeployment(drawdownPct, settings)
                    deploy
                        .setScale(0, RoundingMode.HALF_UP)
                        .shouldBeEqualComparingTo(tableDeployPct)
                }
            }
        }

        // CQ-9-1: ALGORITHM.md MaxDD=30% conservative exponent 2.0 table
        // | DD 1.5% → 0.25% | 7.5% → 6.25% | 15% → 25% | 22.5% → 56% | 30% → 100% |
        "testCalculateFiatDeployment_ConservativeExponentTwo_AlgorithmTable" {
            runTest {
                val settings = TestFixtures.settings(
                    dryRun = false,
                    loopDelaySeconds = 60L,
                    fiatMaxDrawdown = 30.0,
                    fiatDeploymentExponent = 2.0,
                )
                listOf(
                    BigDecimal("1.5") to BigDecimal("0.2500"),
                    BigDecimal("7.5") to BigDecimal("6.2500"),
                    BigDecimal("15") to BigDecimal("25.0000"),
                    BigDecimal("22.5") to BigDecimal("56.2500"),
                    BigDecimal("30") to BigDecimal("100.0000"),
                ).forEach { (drawdownPct, expectedDeployPct) ->
                    portfolioAnalyzer.calculateFiatDeployment(drawdownPct, settings)
                        .shouldBeEqualComparingTo(expectedDeployPct)
                }
                // Documented integer Deploy% (ALGORITHM rounds 56.25% → 56 for display).
                portfolioAnalyzer.calculateFiatDeployment(BigDecimal("22.5"), settings)
                    .setScale(0, RoundingMode.HALF_UP)
                    .shouldBeEqualComparingTo(BigDecimal("56"))
            }
        }

        "testCalculateFiatDeployment_AggressiveExponentHalf_At25PercentOfMaxIs50" {
            runTest {
                // MaxDD 30%, DD 7.5% (25% of max) → (0.25)^0.5 * 100 = 50% exactly
                val settings = TestFixtures.settings(
                    dryRun = false,
                    loopDelaySeconds = 60L,
                    fiatMaxDrawdown = 30.0,
                    fiatDeploymentExponent = 0.5,
                )
                portfolioAnalyzer.calculateFiatDeployment(
                    BigDecimal("7.5"),
                    settings,
                ).shouldBeEqualComparingTo(BigDecimal("50.0"))
            }
        }

        "testCalculateFiatDeployment_AggressiveExponentHalf_AtMaxDrawdownIs100" {
            runTest {
                val settings = TestFixtures.settings(
                    dryRun = false,
                    loopDelaySeconds = 60L,
                    fiatMaxDrawdown = 30.0,
                    fiatDeploymentExponent = 0.5,
                )
                portfolioAnalyzer.calculateFiatDeployment(
                    BigDecimal("30.0"),
                    settings,
                ).shouldBeEqualComparingTo(BigDecimal("100.0"))
            }
        }
    }
}
